app/services/catalog/copy_portfolio_item.rb

```diff
@@ -0,0 +1,41 @@
+module Catalog
+  class CopyPortfolioItem
+    attr_reader :new_portfolio_item
+
+    def initialize(params)
+      @portfolio_item = PortfolioItem.find(params[:portfolio_item_id])
+      @to_portfolio = Portfolio.find_by!(:id => (params[:portfolio_id] || @portfolio_item.portfolio_id))
+      portfolios_write_check
+    end
+
+    def process
+      @new_portfolio_item = make_copy
+      @to_portfolio.portfolio_items << @new_portfolio_item
+
+      self
+    end
+
+    private
+
+    def make_copy
+      new_portfolio_item = @portfolio_item.dup
+
+      if new_portfolio_item.portfolio_id == @to_portfolio.id
+        new_portfolio_item.name = "Copy of " + new_portfolio_item.name
+      end
+
+      new_portfolio_item.save
+      new_portfolio_item
+    end
+
+    def portfolios_write_check
+      return unless RBAC::Access.enabled?
```

Comment:
Is there anyway this check can happen internal to the RBAC classes themselves?

Reply:
After a refactor it's all contained in the RBAC mixin.